### ISSUE TITLE
Allow updating subsets of a target

### ISSUE THREAD

**Issue by DarkerStar** [build2/build2#81]
(Following from a discussion on Slack...)

There is currently no way to request that the build tool compiles only a single translation unit, in practical project setups.

## Demonstration

In an empty directory, create four files:

`buildfile`:

```
using cxx

exe{prog}: cxx{*}
```

`main.cxx`:

```c++
#include "foo.hxx"

auto main() -> int {}
```

`foo.hxx`

```c++
#ifndef INC_foo
#define INC_foo

#endif // include guard
```

`foo.cxx`

```c++
#include "foo.hxx"
```

A simple update compiles both `main.cxx` and `foo.cxx`, and links them into `prog`, as expected:

```
$ b
c++ cxx{main}       
c++ cxx{foo}
ld exe{prog}
$ 
```

What I would like to do (for example) is *just* the second operation, compiling `foo.cxx`, *without* also compiling `main.cxx` or linking `prog`. This is very handy in larger projects when refactoring a widely used component... what should be a rapid edit-compile turnaround with a single translation unit becomes intolerably slow if the *entire project* has to be rebuilt every time.

With the project set up as is, it is impossible to request compiling just `foo.cxx` with `b obj{foo}` or `b obje{foo}`, because those targets don't exist in the buildfile - they are constructed implicitly only when building the dependency tree for `exe{prog}` (which *is* explicitly specified).

A "fix" for this is explicitly defining the targets in the `buildfile`:

```
using cxx

exe{prog}: cxx{*}

obje{foo}: cxx{foo}
```

So now you can do `b obje{foo}`. However, you would need to do this for every C++ translation unit in the project (the above still won't work for `b obje{main}`, for example). (It also won't work for the simpler `obj{foo}` target.)

## Discussion

The problem, as alluded to in [an email discussion](https://lists.build2.org/archives/users/2020-February/000741.html), is that implicit intermediate targets just don't exist until the target that depends on them is instantiated. There is simply no `obje{foo}` until the build system constructing the dependency tree for `exe{prog}` realizes it needs one to transform `cxx{foo}` into something `exe{prog}` can use.

That means that to find arbitrary intermediate objects, one would need to construct the entire dependency tree... for every target. That's not a *terrible* idea - `b clean` needs to do that anyway.

But perhaps it could be possible to specify a target... and then request that only a subset of that target's dependencies get updated.

For example, to build the target program in the above example, you would do `b exe{prog}`. Internally the build tool must be doing something like:

```
exe{prog}
: cxx{**}
⇩
exe{prog}
: obj{**}:cxx{**}
⇩
exe{prog}
: obj{**}:obje{**}:cxx{**}
⇩
exe{prog}
: obj{main}:obje{main}:cxx{main}
: obj{foo}:obje{foo}:cxx{foo}
⇩
exe{prog}
: obj{main}
: : obje{main}
: : : cxx{main}
: : : hxx{foo}
: obj{foo}
: : obje{foo}
: : : cxx{foo}
: : : hxx{foo}
```

And then it runs down the tree doing whatever the `perform` operation is - usually `update`.

Since this tree is being built anyways, it would be nice if we could single out parts of it.

For example, `b update: exe{prog}[obje{foo}]` (using hypothetical syntax that's open to bikeshedding) would *only* update the `obje{foo}` target in the tree above... which would require updating `cxx{foo}` and `hxx{foo}` (which in some cases might in turn involve updating something like `in{foo}`, and so on). The net result is that (in this case) only `foo.cxx` gets compiled. `main.cxx` is not considered, and `prog` is not linked.

Since `update` is the default option, and (in this project) `exe{prog}` is the default target, and the `obje{???}` target class is only dependency of the group `obj{???}`, that could even be simplified to: `b [obj{foo}]`.

With the ability to select subsets of a target, one could do things like:

* `b src/@out/exe{prog}[obj{foo}]`: compile only `foo.cxx` for executable target `prog`, specifying `out_base` and `src_base`.
* `b [obj{foo}]`: same as above, assuming `out_base` and `src_base` are configured, and `exe{prog}` is the default target.
* `b exe{prog}[obj{**}]`: compile everything for `prog`, but don't link the exe. Handy if you know it won't link anyway, because maybe the system libs are ABI-incompatible (like if you're using clang with `-std=libc++` on Linux).
* `b '{clean update}([obj{foo}])'`: recompile just `foo.cxx`. Handy if something outside of the build system's ken requires recompiling just one or two files, rather than recompiling *everything* - you could just recompile only those files that need it, then do `b` to relink.
* `b [bmi{foo}]`: update a module interface.

Note that if you don't specify a target to build a subset of, it would be cool if it just uses the default target... which might be `./` (from the target root)... which might require building the dependency tree for *the entire project*. But that's fine; it's what the user asked for, and it's a reasonable, safe default. If they want more efficiency, they can specify a target.

(Once it's possible to specify subsets of a target, another handy capability to have would be "minimal updates". For example, if I specify `b minimal-update: [obj{foo}]` it would update *just* that single target, and not its dependencies. It would *just* compile `foo.o`. It would *not* update any of `foo.o`'s dependencies... so, for example, if `foo.cxx` included `config.hxx`, which is generated from `config.hxx.in`, that does not happen. If `config.hxx` exists, even if it's out of date, it's used. (If it doesn't exist, that's a compile error, but that's fine... you got what you asked for.) But let's consider the ability to specify target subsets, first, and shelve "minimal updates" for possible consideration after.)

**Comment by boris-kolpackov** [comment]
Thanks!

My comment from the Slack discussion:

> I also thought that if only we could match the relevant declared target, then that would trigger the creation of the implicit ones which we could then use. I also first thought of specifying the declared target explicitly, but then realized that we could just match `./` by default and reasonably expect it to pull all the implicit ones. There could be a bit of overhead but I think it will be a lot user-friendlier. We can at least start there and see how it goes.

Also, we will need to have some sensible default member selection for `obj{foo}` (currently one cannot build `obj{}` directly and has to select one of its `obj[ase]{}` members). We could probably use the link order (for example `config.bin.exe.lib`) to choose between `obja{}` and `objs{}`.

**Comment by DarkerStar** [comment]
> Also, we will need to have some sensible default member selection for `obj{foo}` (currently one cannot build `obj{}` directly and has to select one of its `obj[ase]{}` members). We could probably use the link order (for example `config.bin.exe.lib`) to choose between `obja{}` and `objs{}`.

Ah, yeah, I've been mostly working with exe projects (using build2 for every little utility I need for every little task to really get used to it - but come to think of it, I should really move my personal code library to build2), so I didn't think about static-vs-shared libs.

My instinct would be to interpret `obj{}` as *everything* it wraps. So specifying `obj{foo}` would try updating `obje{foo}`, `obja{foo}`, `objs{foo}`, and anything else it might possibly mean now or in the future. So the build tool doesn't need to choose between `obja{}` and `objs{}`... it assumes both, and all other possibilities, and discards any that don't apply - that is, it only considers the `obj?{}` results that actually exist as dependencies of the given target. For an exe target `exe{prog}`, given `exe{prog}[obj{foo}]`, it could try all the possibilities, realize that most (like `obja{foo}`) aren't actually dependencies of `exe{prog}`, and end up with only `obje{foo}`... and that's what it updates. For a library project where the root target builds both `libs{thing}` and `liba{thing}`, `b [obj{foo}]` could update `obja{foo}` *and* `objs{foo}`... and if it was only building shared libs, then `b [obj{foo}]` would only update `objs{foo}`, because `obja{foo}` wasn't a dependency of `./`.

And of course, the door is open to be specific if the programmer really wants to be, by explicitly saying `[obje{foo}]` or `[objs{foo}]` or `libs{thing}[obj{foo}]` or whatever if they please.

**Comment by boris-kolpackov** [comment]
Yes, that's actually an even better and simpler idea. And you are right, one will be able to specify one of the member directly to avoid building the other member (in most case it will either be `obje{}` or `obja{}` and `objs{}`).
